### ISSUE TITLE
No warnings: declaration makes variable indeterminate

### ISSUE THREAD

**Issue by ch3root** [TrustInSoft/tis-interpreter#79]
Source code:
```c
#include <stdio.h>

int main()
{
  goto step1;

 step2:;
  int a; // (1)
  //int a = a; // (2)
  printf("a = %d\n", a);
  return 0;
  
 step1:
  a = 1;
  goto step2;
}
```

tis-interpreter (e0dfac063885a1740751c15b8dec9a0a8e5191d7) output:
```
[value] Analyzing a complete application starting at main
[value] Computing initial state
[value] Initial state computed

a = 1

[value] done for function main
```

When printf is reached the value of `a` is indeterminate. There should be a warning.

C11, 6.2.4p6: "If an initialization is specified for the object, [...]; otherwise, the value becomes indeterminate each time the declaration is reached."

I think replacing (1) by (2) permits `a` to retain its value.

gcc 7 w/ubsan output:
```
a = 1
```

clang 3.9 w/ubsan output:
```
init-self-goto.c:10:22: warning: variable 'a' is uninitialized when used here [-Wuninitialized]
  printf("a = %d\n", a);
                     ^
init-self-goto.c:8:8: note: initialize the variable 'a' to silence this warning
  int a; // (1)
       ^
        = 0
1 warning generated.
a = 1
```

**Comment by pascal-cuoq** [comment]
I think this can be fixed. Currently we set `a` to uninitialized each time the block that contains it is entered (if I remember correctly, this block is the scope of `a`). We can make it uninitialized again each time the declaration is encountered.

With that system, the `int a = a;` case may not cause a warning, but leave `a` uninitialized. I guess I never know how to handle uninitialized memory anyway.

**Comment by ch3root** [comment]
The scope is where `a` is visible, I think you mean lifetime. For easier reference, full paragraphs 5 and 6 from subclause 6.2.4 from C11, splitted by sentence, and numbered:

(5.1) An object whose identifier is declared with no linkage and without the storage-class specifier static has automatic storage duration, as do some compound literals.
(5.2) The result of attempting to indirectly access an object with automatic storage duration from a thread other than the one with which the object is associated is implementation-defined.

(6.1) For such an object that does not have a variable length array type, its lifetime extends from entry into the block with which it is associated until execution of that block ends in any way.
(6.2) (Entering an enclosed block or calling a function suspends, but does not end, execution of the current block.)
(6.3) If the block is entered recursively, a new instance of the object is created each time.
(6.4) The initial value of the object is indeterminate.
(6.5) If an initialization is specified for the object, it is performed each time the declaration or compound literal is reached in the execution of the block; otherwise, the value becomes indeterminate each time the declaration is reached.

According to (6.1), the lifetime of `a` indeed start when its block is entered.
According to (6.4), at this moment `a` should indeed be set to uninitialized.
(6.5) seems to mean that `a` should be reset to uninitialized each time its declaration is encountered but only if this declaration lacks initializer. So `int a = a` will keep the value of `a` intact whether it's uninitialized or not. In my example `a` would be already set to 1.

**Comment by pascal-cuoq** [comment]
Yes I have a big problem with scope, lifetime and linkage. Always assume I meant the other one.

**Comment by ch3root** [comment]
I prefer to have the definitions of these terms in front of me when I write about them:-) I'm starting to think a cheatsheet is needed...

Anyway, another relevant quote. C11, 6.8p3: "The initializers of objects that have automatic storage duration, and the variable length array declarators of ordinary identifiers with block scope, are evaluated and the values are stored in the objects (including storing an indeterminate value in objects without an initializer) each time the declaration is reached in the order of execution, as if it were a statement, and within each declaration in the order that declarators appear."
Especially the part in parentheses.
